File: LR0/src/GUI/GUI.java
package GUI;

import CycleList.CycleList;
import Factory.UserFactory;
import Types.Users.UserType;

import java.awt.*;
import java.awt.event.*;
import java.io.FileNotFoundException;
import javax.swing.*;
import javax.xml.stream.XMLStreamException;

public class GUI extends JPanel {
    private JButton findBtn;
    private JButton delBtn;
    private JButton insertByIdBtn;
    private JButton insertBtn;
    private JButton sortBtn;
    private JButton saveBtn;
    private JButton loadBtn;
    private JButton clearBtn;
    private JLabel findLabel;
    private JLabel delLabel;
    private JLabel typeLabel;
    private JLabel insertLabel;
    private JLabel createLabel;
    private JLabel sortLabel;
    private JTextArea outTextField;
    private JTextField findByIdField;
    private JTextField delByIdField;
    private JTextField insertByIdField;
    private JComboBox typeList;
    public UserFactory userFactory;
    public UserType userType;
    public CycleList cycleList;
    public String defaultType;

    public final String FILE_NAME_INTEGER = "CycleListInteger.dat";
    public final String FILE_NAME_FRACTION = "CycleListFraction.dat";

    public GUI() {
        defaultType = "Integer";
        userFactory = new UserFactory();
        userType = userFactory.getBuilderByName(defaultType);
        cycleList = new CycleList();
        JComboBox<String> factoryListItems = new JComboBox<>(UserFactory.getTypeNameList().toArray(new String[0]));

        findBtn = new JButton("Найти");
        delBtn = new JButton("Удалить");
        insertByIdBtn = new JButton("Вставить");
        insertBtn = new JButton("Сгенерировать");
        sortBtn = new JButton("Сортировать");
        saveBtn = new JButton("Сохранить");
        loadBtn = new JButton("Загрузить");
        clearBtn = new JButton("Очистить");
        outTextField = new JTextArea(5, 5);
        typeList = factoryListItems;
        findByIdField = new JTextField(5);
        delByIdField = new JTextField(5);
        insertByIdField = new JTextField(5);
        findLabel = new JLabel("Поиск по индексу");
        delLabel = new JLabel("Удалить по индексу");
        typeLabel = new JLabel("Тип данных");
        insertLabel = new JLabel("Вставка по индексу");
        createLabel = new JLabel("Сгенерировать данные");
        sortLabel = new JLabel("Сортировка");

        outTextField.setEnabled(true);
        outTextField.setFont(new Font("Arial", Font.BOLD, 14));

        setPreferredSize(new Dimension(600, 563));
        setLayout(null);

        add(findBtn);
        add(delBtn);
        add(insertByIdBtn);
        add(outTextField);
        add(typeList);
        add(findByIdField);
        add(findLabel);
        add(delByIdField);
        add(delLabel);
        add(typeLabel);
        add(insertLabel);
        add(insertByIdField);
        add(insertBtn);
        add(createLabel);
        add(sortLabel);
        add(sortBtn);
        add(saveBtn);
        add(loadBtn);
        add(clearBtn);

        findBtn.setBounds(25, 135, 150, 25);
        delBtn.setBounds(25, 220, 150, 25);
        insertByIdBtn.setBounds(25, 315, 150, 25);
        outTextField.setBounds(190, 20, 345, 525);
        typeList.setBounds(25, 50, 150, 25);
        findByIdField.setBounds(25, 105, 150, 25);
        findLabel.setBounds(25, 85, 150, 20);
        delByIdField.setBounds(25, 190, 150, 25);
        delLabel.setBounds(25, 165, 150, 25);
        typeLabel.setBounds(25, 20, 150, 25);
        insertLabel.setBounds(25, 255, 150, 25);
        insertByIdField.setBounds(25, 280, 150, 25);
        insertBtn.setBounds(25, 435, 150, 20);
        createLabel.setBounds(25, 405, 145, 25);
        sortLabel.setBounds(25, 350, 100, 25);
        sortBtn.setBounds(25, 375, 150, 25);
        saveBtn.setBounds(25, 465, 150, 20);
        loadBtn.setBounds(25, 490, 150, 20);
        clearBtn.setBounds(25, 520, 150, 25);

        findBtn.addActionListener(e -> findNodeById());
        delBtn.addActionListener(e -> deleteNodeById());
        insertByIdBtn.addActionListener(e -> {
            try {
                addNodeById();
            } catch (InstantiationException ex) {
                throw new RuntimeException(ex);
            }
        });
        insertBtn.addActionListener(e -> {
            try {
                addNode();
            } catch (InstantiationException ex) {
                throw new RuntimeException(ex);
            }
        });
        sortBtn.addActionListener(e -> sortList());
        saveBtn.addActionListener(e -> saveList());
        loadBtn.addActionListener(e -> {
            try {
                loadList();
            } catch (XMLStreamException ex) {
                throw new RuntimeException(ex);
            } catch (FileNotFoundException ex) {
                throw new RuntimeException(ex);
            }
        });
        clearBtn.addActionListener(e -> clearOutTextField());
        typeList.addActionListener(e -> selectTypeList(e));
    }

    public void showMenu() {
        JFrame frame = new JFrame("MainMenu");
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
        frame.getContentPane().add(new GUI());
        frame.pack();
        frame.setVisible(true);
        frame.setResizable(false);
    }


    private void selectTypeList(ActionEvent actionEvent) {
        JComboBox box = (JComboBox) actionEvent.getSource();
        String item = (String) box.getSelectedItem();
        if (defaultType != item) {
            defaultType = item;
            userType = userFactory.getBuilderByName(defaultType);
            cycleList = new CycleList();
            setTextOnOutTextField();
        }
    }

    private void addNode() throws InstantiationException {
        cycleList.add(userType.create());
        setTextOnOutTextField();

    }

    private void addNodeById() throws InstantiationException {
        if (insertByIdField.getText().isEmpty()) {
            JOptionPane.showMessageDialog(null, "Пустое поле! Введите значение!");
            return;
        }
        cycleList.add(userType.create(), Integer.parseInt(insertByIdField.getText()));
        setTextOnOutTextField();
    }

    private void deleteNodeById() {
        if (delByIdField.getText().isEmpty()) {
            JOptionPane.showMessageDialog(null, "Пустое поле! Введите значение!");
            return;
        }
        cycleList.remove(Integer.parseInt(delByIdField.getText()));
        setTextOnOutTextField();
    }

    private void sortList() {
        cycleList.sort(userType.getTypeComparator());
        setTextOnOutTextField();
    }

    private void findNodeById() {
        if (findByIdField.getText().isEmpty()) {
            JOptionPane.showMessageDialog(null, "Пустое поле! Введите значение!");
            return;
        }
        JOptionPane.showMessageDialog(null, "Найденное значение: " +
                cycleList.getByIndex(Integer.parseInt(findByIdField.getText())));
    }

    private void saveList() {
        if (defaultType == "Integer") {
            cycleList.save(userType, FILE_NAME_INTEGER);
            JOptionPane.showMessageDialog(null, "Список успешно сохранен в \"" + FILE_NAME_INTEGER + "\"!");
        } else {
            cycleList.save(userType, FILE_NAME_FRACTION);
            JOptionPane.showMessageDialog(null, "Список успешно сохранен в \"" + FILE_NAME_FRACTION + "\"!");
        }

    }

    private void loadList() throws XMLStreamException, FileNotFoundException {
        if (defaultType == "Integer") {
            cycleList.load(userType, FILE_NAME_INTEGER);
            JOptionPane.showMessageDialog(null, "Список успешно загружен!");
            setTextOnOutTextField();
        } else {
            cycleList.load(userType, FILE_NAME_FRACTION);
            JOptionPane.showMessageDialog(null, "Список успешно загружен!");
            setTextOnOutTextField();
        }
    }

    private void clearOutTextField() {
        cycleList = new CycleList();
        setTextOnOutTextField();
    }

    private void setTextOnOutTextField() {
        outTextField.setText(cycleList.toString());
    }
}

File: LR2/src/main/kotlin/gui/GUI.kt
package gui

import cycle_list.CycleList
import factory.UserFactory
import types.users.UserType
import java.awt.Dimension
import java.awt.Font
import java.awt.event.ActionEvent
import java.io.FileNotFoundException
import javax.swing.*
import javax.xml.stream.XMLStreamException

class GUI : JPanel() {
    private val findBtn: JButton
    private val delBtn: JButton
    private val insertByIdBtn: JButton
    private val insertBtn: JButton
    private val sortBtn: JButton
    private val saveBtn: JButton
    private val loadBtn: JButton
    private val clearBtn: JButton
    private val findLabel: JLabel
    private val delLabel: JLabel
    private val typeLabel: JLabel
    private val insertLabel: JLabel
    private val createLabel: JLabel
    private val sortLabel: JLabel
    private val outTextField: JTextArea
    private val findByIdField: JTextField
    private val delByIdField: JTextField
    private val insertByIdField: JTextField
    private val typeList: JComboBox<*>
    var userFactory: UserFactory
    var userType: UserType?
    var cycleList: CycleList
    var defaultType = "Integer"
    val FILE_NAME_INTEGER = "CycleListInteger.dat"
    val FILE_NAME_FRACTION = "CycleListFraction.dat"

    init {
        userFactory = UserFactory()
        userType = userFactory.getBuilderByName(defaultType)
        cycleList = CycleList()

        val typeNameList = userFactory.getTypeNameList()
        val factoryListItemsItems = arrayOfNulls<String>(typeNameList.size)
        for (i in typeNameList.indices) {
            factoryListItemsItems[i] = typeNameList[i]
        }
        val factoryListItems = JComboBox<String>(factoryListItemsItems)
        factoryListItems.setSelectedItem(defaultType)

        findBtn = JButton("Найти")
        delBtn = JButton("Удалить")
        insertByIdBtn = JButton("Вставить")
        insertBtn = JButton("Сгенерировать")
        sortBtn = JButton("Сортировать")
        saveBtn = JButton("Сохранить")
        loadBtn = JButton("Загрузить")
        clearBtn = JButton("Очистить")
        outTextField = JTextArea(5, 5)
        typeList = factoryListItems
        findByIdField = JTextField(5)
        delByIdField = JTextField(5)
        insertByIdField = JTextField(5)
        findLabel = JLabel("Поиск по индексу")
        delLabel = JLabel("Удалить по индексу")
        typeLabel = JLabel("Тип данных")
        insertLabel = JLabel("Вставка по индексу")
        createLabel = JLabel("Сгенерировать данные")
        sortLabel = JLabel("Сортировка")
        outTextField.setEnabled(true)
        outTextField.setFont(Font("Arial", Font.BOLD, 14))
        preferredSize = Dimension(600, 563)
        setLayout(null)
        add(findBtn)
        add(delBtn)
        add(insertByIdBtn)
        add(outTextField)
        add(typeList)
        add(findByIdField)
        add(findLabel)
        add(delByIdField)
        add(delLabel)
        add(typeLabel)
        add(insertLabel)
        add(insertByIdField)
        add(insertBtn)
        add(createLabel)
        add(sortLabel)
        add(sortBtn)
        add(saveBtn)
        add(loadBtn)
        add(clearBtn)
        findBtn.setBounds(25, 135, 150, 25)
        delBtn.setBounds(25, 220, 150, 25)
        insertByIdBtn.setBounds(25, 315, 150, 25)
        outTextField.setBounds(190, 20, 345, 525)
        typeList.setBounds(25, 50, 150, 25)
        findByIdField.setBounds(25, 105, 150, 25)
        findLabel.setBounds(25, 85, 150, 20)
        delByIdField.setBounds(25, 190, 150, 25)
        delLabel.setBounds(25, 165, 150, 25)
        typeLabel.setBounds(25, 20, 150, 25)
        insertLabel.setBounds(25, 255, 150, 25)
        insertByIdField.setBounds(25, 280, 150, 25)
        insertBtn.setBounds(25, 435, 150, 20)
        createLabel.setBounds(25, 405, 145, 25)
        sortLabel.setBounds(25, 350, 100, 25)
        sortBtn.setBounds(25, 375, 150, 25)
        saveBtn.setBounds(25, 465, 150, 20)
        loadBtn.setBounds(25, 490, 150, 20)
        clearBtn.setBounds(25, 520, 150, 25)
        findBtn.addActionListener { e: ActionEvent? -> findNodeById() }
        delBtn.addActionListener { e: ActionEvent? -> deleteNodeById() }
        insertByIdBtn.addActionListener { e: ActionEvent? ->
            try {
                addNodeById()
            } catch (ex: InstantiationException) {
                throw RuntimeException(ex)
            }
        }
        insertBtn.addActionListener { e: ActionEvent? ->
            try {
                addNode()
            } catch (ex: InstantiationException) {
                throw RuntimeException(ex)
            }
        }
        sortBtn.addActionListener { e: ActionEvent? -> sortList() }
        saveBtn.addActionListener { e: ActionEvent? -> saveList() }
        loadBtn.addActionListener { e: ActionEvent? ->
            try {
                loadList()
            } catch (ex: XMLStreamException) {
                throw RuntimeException(ex)
            } catch (ex: FileNotFoundException) {
                throw RuntimeException(ex)
            }
        }
        clearBtn.addActionListener { e: ActionEvent? -> clearOutTextField() }
        typeList.addActionListener { e: ActionEvent -> selectTypeList(e) }
    }

    fun showMenu() {
        val frame = JFrame("MainMenu")
        frame.setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
        frame.contentPane.add(GUI())
        frame.pack()
        frame.isVisible = true
        frame.setResizable(false)
    }

    private fun selectTypeList(actionEvent: ActionEvent) {
        val box = actionEvent.source as JComboBox<*>
        val item = box.selectedItem as String
        if (defaultType !== item) {
            defaultType = item
            userType = userFactory.getBuilderByName(defaultType)
            cycleList = CycleList()
            setTextOnOutTextField()
        }
    }

    @Throws(InstantiationException::class)
    private fun addNode() {
        cycleList.add(userType!!.create())
        setTextOnOutTextField()
    }

    @Throws(InstantiationException::class)
    private fun addNodeById() {
        if (insertByIdField.getText().isEmpty()) {
            JOptionPane.showMessageDialog(null, "Пустое поле! Введите значение!")
            return
        }
        cycleList.add(userType!!.create(), insertByIdField.getText().toInt())
        setTextOnOutTextField()
    }

    private fun deleteNodeById() {
        if (delByIdField.getText().isEmpty()) {
            JOptionPane.showMessageDialog(null, "Пустое поле! Введите значение!")
            return
        }
        cycleList.remove(delByIdField.getText().toInt())
        setTextOnOutTextField()
    }

    private fun sortList() {
        cycleList.sort(userType!!.typeComparator)
        setTextOnOutTextField()
    }

    private fun findNodeById() {
        if (findByIdField.getText().isEmpty()) {
            JOptionPane.showMessageDialog(null, "Пустое поле! Введите значение!")
            return
        }
        JOptionPane.showMessageDialog(
            null, "Найденное значение: " +
                    cycleList.getByIndex(findByIdField.getText().toInt())
        )
    }

    private fun saveList() {
        if (defaultType === "Integer") {
            cycleList.save(userType!!, FILE_NAME_INTEGER)
            JOptionPane.showMessageDialog(null, "Список успешно сохранен в \"$FILE_NAME_INTEGER\"!")
        } else {
            cycleList.save(userType!!, FILE_NAME_FRACTION)
            JOptionPane.showMessageDialog(null, "Список успешно сохранен в \"$FILE_NAME_FRACTION\"!")
        }
    }

    @Throws(XMLStreamException::class, FileNotFoundException::class)
    private fun loadList() {
        if (defaultType === "Integer") {
            cycleList.load(userType!!, FILE_NAME_INTEGER)
            JOptionPane.showMessageDialog(null, "Список успешно загружен!")
            setTextOnOutTextField()
        } else {
            cycleList.load(userType!!, FILE_NAME_FRACTION)
            JOptionPane.showMessageDialog(null, "Список успешно загружен!")
            setTextOnOutTextField()
        }
    }

    private fun clearOutTextField() {
        cycleList = CycleList()
        setTextOnOutTextField()
    }

    private fun setTextOnOutTextField() {
        outTextField.text = cycleList.toString()
    }
}
